source/extensions/filters/network/ratelimit/ratelimit.h

```diff
@@ -83,13 +85,44 @@ class Filter : public Network::ReadFilter,
   void initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callbacks) override {
     filter_callbacks_ = &callbacks;
     filter_callbacks_->connection().addConnectionCallbacks(*this);
+    applySubstitutionFormatter(config_->descriptors(),
+                               filter_callbacks_->connection().streamInfo());
+  }
+
+  void applySubstitutionFormatter(std::vector<RateLimit::Descriptor> original_descriptors,
+                                  StreamInfo::StreamInfo& stream_info) {
+
+    std::vector<RateLimit::Descriptor> dynamicDescriptors = std::vector<RateLimit::Descriptor>();
+    for (const RateLimit::Descriptor& descriptor : original_descriptors) {
+      RateLimit::Descriptor new_descriptor;
+      for (const RateLimit::DescriptorEntry& descriptorEntry : descriptor.entries_) {
+        std::string value = descriptorEntry.value_;
+        value = formatValue(value, stream_info);
+        new_descriptor.entries_.push_back({descriptorEntry.key_, value});
+      }
+      dynamicDescriptors.push_back(new_descriptor);
+    }
+    filter_descriptors_ = dynamicDescriptors;
+  }
+
+  std::string formatValue(std::string body, StreamInfo::StreamInfo& stream_info) {
+    Http::RequestHeaderMapPtr request_headers = Http::RequestHeaderMapImpl::create();
+    Http::ResponseHeaderMapPtr response_headers = Http::ResponseHeaderMapImpl::create();
+    Http::ResponseTrailerMapPtr response_trailers = Http::ResponseTrailerMapImpl::create();
```

Comment:
For perf reasons I would create some static headers and trailers for use to avoid construction every time you need to do this. These can live in the config object also per my below comment so that when you do the formatting through the config you can just use the static values.